pkg/planner/core/plan_cache.go

```diff
@@ -516,17 +533,25 @@ func rebuildRange(p Plan) error {
 		}
 		return nil
 	case *BatchPointGetPlan:
+		if x.TblInfo.GetPartitionInfo() != nil && fixcontrol.GetBoolWithDefault(sctx.GetSessionVars().OptimizerFixControl, fixcontrol.Fix33031, false) {
+			sctx.GetSessionVars().StmtCtx.UseCache = false
+			return errors.New("Fix33031 fix-control set and partitioned table in cached Batch Point Get plan")
+		}
 		// if access condition is not nil, which means it's a point get generated by cbo.
 		if x.AccessConditions != nil {
 			if x.IndexInfo != nil {
 				ranges, err := ranger.DetachCondAndBuildRangeForIndex(x.ctx, x.AccessConditions, x.IdxCols, x.IdxColLens, 0)
 				if err != nil {
 					return err
 				}
+				if len(ranges.Ranges) != len(x.IndexValues) {
+					// Previous run may have removed duplicates or non matching partitions
```

Comment:
In this case, we'd better generate a new plan rather than reusing the old one for safety.